### DISCUSSION TITLE
Translators: change in website translation workflow

### DISCUSSION THREAD

**Discussion by ignotus666** [jamulussoftware/jamulus#1725] [Documentation]
@jamulussoftware/translators:

As those of you who have translated website documentation know, for the last couple of releases we've been using a system whereby several branches (changes, translate, release) exist to keep track of changes and bundle them together in a single PR for translators, who then (if not starting from scratch) have to trawl through the documentation in their language to locate and edit the relevant parts. While functional, this system is not ideal as the existence of several branches that have to interact with each other in a certain way tends to lead to confusion, and the translation process itself is far from optimal as it involves spending a lot of time just locating where changes have happened.

Jamulus team member @hoffie recently brought the [po4a](https://po4a.org/) tool to our attention to see if it might help with the translation workflow. I've been looking into it and it looks like it could be feasible to use it to make the translation and documentation maintenance process (IMHO) simpler and more user-friendly. The transition involves converting existing translations into the .po format, a fairly tedious process that is nonetheless almost complete. These .po files would be what translators would work with from now on.

PO files contain the original text separated into strings (segments), along with the translation (if it exists) for each segment. There are several editors you can use, such as poedit, Lokalize or my preferred tool, OmegaT. The process is similar to what you get with QtLinguist, for those of you who have done Jamulus app translations. So basically, instead of waiting for the PR with all the changes to the original docs and directly editing .md files, translators simply have to work on the .po files for their language. Whenever a change is made to an original document in English, a script is run that propagates those changes to all the .po files, which will then show them as untranslated strings. Once all strings have been translated, you submit the .po file(s) via a PR and once merged, another script converts them back into the translated version of the .md file, ready to be used for the website. These scripts do not have to be run by the translator; hopefully it will be possible to run them via a GH action, or in the worst case, a team member can run them.

In short, the workflow would be as follows.

**For translators**:
- Fork website repo. There would not be multiple branches to worry about - just the one.
- Translate .po files (.md files must not be touched!).
- Submit PR.
- Done.

**For website content maintainers**:
- Edit/create content (.md files). Run script*.
- Received PR with translated .po files? Run script*.
- Done.
- No need to set up several interacting branches with all the complexity it brings.

*Could be run via a GH action

I've been setting up the scripts and directory structure to make the transition, and by next week I hope to set up a branch to test the waters, but before doing the switch we thought it'd be good to see what others might have to say. Does anyone see any drawbacks, have a negative experience of such a workflow? Do you have any suggestions or experience using po4a? I'm completely new to it and while I think I can cobble together a functional setup, any help, criticism or suggestions are more than welcome!

**Comment by hoffie** [comment]
This sounds really good, thanks for moving this forward! :)

Possible additional advantages:
- It might become possible to determine if and by how much specific translations are outdated (which is not easily possible today besides comparing time stamps).

Possible disadvantage:
- While .po is a text-based format, as far as I know, it might be a bit harder to do diffs around them. The need for diffing should vanish for most cases, though.

Note: I assume we'd still need two branches though -- one which resembles what's live on the website (current `release`) and one which includes what's up for the next release. The latter would be what translators would usually interact with. I suspect that no additional `translate3_y_z` branch will be necessary anymore.

**Reply by ignotus666** [reply]
It's possible to set a threshold below which the .md file doesn't get created from the .po. I've set it provisionally at 80%, so if a translation gets outdated by more than 20%, it gets pulled.

Yes, you're right that we'd probably still have 2 branches - but translators (and maintainers) need only be concerned with one, which receives all changes, both translations and documentation updates. The complexity should be vastly reduced.

**Reply by gilgongo** [reply]
> if a translation gets outdated by more than 20%, it gets pulled

I wonder what the practical effect of such a threshold system on the website would be? If we had (for example) 80% on the website, then the untranslated 20% would likely contain new information for a release in many cases. So is it better to (unknowingly) get an outdated translation or none at all, I wonder? I guess is may be less of an issue for the app perhaps if 20% of strings you see are in English.

**Reply by ignotus666** [reply]
It's an arbitrary threshold that can be changed at any time - if you set it at 95% just a few missing strings will prevent its publication. 100% wouldn't be practical as the occasional string like numbers or some links are left untranslated. It's still better than what we have though. Yesterday I attempted to align the few existing files in Italian and they are so outdated I was barely able to use any of their content - but those pages are active right now on the website.

**Reply by gilgongo** [reply]
Yes, I guess no system will be perfect, and as you say what we have now it not good.

**Reply by hoffie** [reply]
Yes. I also assume that no system will take the decision from us regarding what to do with partial translations. Having a system which is at least configurable to do what we decide upon sounds like the best possible approach.

**Reply by ignotus666** [reply]
This came up on the 3_8_0 release discussion thread, but I thought I'd address it here to avoid going OT there.

To clarify: when a doc in EN gets updated, and that update doesn't get translated, the modified fragments will be displayed in English in the translated text. This means that outdated text will never be shown, in any language, if it's been updated in English. In practice, the threshold determines the amount of EN text sprinkled into a translation we're prepared to tolerate before pulling it entirely.

**Reply by gilgongo** [reply]
Just so I understand - p4oa treats a "fragment" here as a paragraph, yes?

**Reply by ignotus666** [reply]
Yes. If something changes in a paragraph, the whole thing is treated as a unit (a "segment"). Then, within segments, editors can suggest "fuzzy matches" where, again, a threshold can be set to use them or not. But that's getting into finer details.

**Reply by gilgongo** [reply]
OK, it's the "fuzzy" stuff I was wondering about in case we made a small but significant change to a segment. But I can't think of any previous examples so maybe cross that bridge later.

**Reply by ignotus666** [reply]
In OmegaT you can choose whether to automatically insert fuzzy matches or not, and if so, above what threshold. Not sure about the other editors though, I'll have to look into it.

**Comment by ignotus666** [comment]
If anyone wants to try out the use of po4a for the website documentation and translation I've created a branch [here](https://github.com/ignotus666/jamuluswebsite/tree/po4a).

The readme in the wiki folder explains a bit about how it would be used. To test it locally (Linux only I'm afraid), clone the branch and install po4a via your distro's package manager or compile the latest version from the [po4a GitHub repo](https://github.com/mquinson/po4a).

Have a go at editing EN .md files and running the scripts, editing .po files and generating the target .md files... and see if it works as expected. I am VERY amateur at this scripting business and I'm sure things can be done more efficiently, so feel free to make suggestions.

**Reply by ignotus666** [reply]
For those interested in doing website translations I've added a guide for using OmegaT to the po4a branch (in the translator-files folder). Poedit and Lokalize can be used, but OmegaT is vastly superior when it comes to the translation work itself, mainly thanks to the much better user interface that offers a clearer view of the translation units. It's not as intuitive to set up though, hence the guide, but once you get used to it I think it makes the job much easier.

**Comment by jujudusud** [comment]
If the worflow is simpler, safer, powerfull ... then we should go for it !

**Reply by ignotus666** [reply]
IMHO OmegaT is by far the best open source computer-aided translation tool out there. In my daily work I have to use a €700 piece of proprietary software that does all kinds of fancy stuff... that I never use. Not to mention crashes, freezes, buggy behaviour, etc. On Linux at least, OmegaT has not crashed/frozen/done weird stuff on me _ever_ over the last 15 years.

**Comment by ann0see** [comment]
Just ran your scripts locally and I'm impressed. It needs rethinking since it's a totally different process, but probably it's worth it. I haven't done much with OmegaT, after having it set up, but it seems to be a very powerful program - and certainly needs some work for newbies ;-).

Could you please document somewhere that your scripts need [rename](https://packages.debian.org/en/buster/rename) and po4a (of course)? In future this process should be automated with GH-Actions (e.g. run scripts on each commit).

I think we need to remove the en- prefix for all files and should move the translator-files folder into the root directory. This might avoid people editing the content in wiki/lang* manually since they don't look into these folders.

**Reply by ignotus666** [reply]
It'll take some adjusting to but I think it's got potential to simplify things quite a lot. The thing about OmegaT is that while it may seem daunting at first, once you've translated a couple of docs you get the hang of it. It might not be worth learning how to use it for a couple of texts, but if you want to work through a large volume like we have here, it's well worth the initial effort.

When people submit translated .po files, the create-all-targets.sh script could be run automatically, but I think we'll want more control over the other one, update-templates.sh. When the time comes we should do some dry runs. There are a few things that will need to be taken into account but I think it's better to look at them while going through the process. 

Sure, I'll document what's needed. I want to get a couple of other things out of the way first.
